mlflow/gateway/providers/mlflow.py

```diff
@@ -0,0 +1,143 @@
+            base_url=self.mlflow_config.mlflow_server_url,
+            path="invocations",
+            payload=self.process_payload(payload, "prompt"),
+        )
+
+        # Example response:
+        # {"predictions": ["hello", "hi", "goodbye"]}
+
+        return completions.ResponsePayload(
+            **{
+                "candidates": self.process_response(resp["predictions"]),
+                "metadata": {
+                    "model": self.config.model.name,
+                    "route_type": self.config.route_type,
+                },
+            }
+        )
+
+    async def chat(self, payload: chat.RequestPayload) -> chat.ResponsePayload:
+        # Example request to MLflow REST API for chat:
+        # {
+        #     "inputs": ["question"],
+        #     "params": ["temperature": 0.2],
+        # }
+
+        payload = self.process_payload(payload, "messages")
+
+        query_count = len(payload["inputs"])
+        if query_count > 1:
+            raise HTTPException(
+                status_code=422,
+                detail="MLflow chat models are only capable of processing a single query at a "
+                f"time. The request submitted consists of {query_count} queries.",
+            )
+
+        payload["inputs"] = [payload["inputs"][0]["content"]]
+
+        resp = await send_request(
+            headers=self.headers,
+            base_url=self.mlflow_config.mlflow_server_url,
+            path="invocations",
+            payload=payload,
+        )
+
+        # Example response:
+        # {"predictions": "answer"}
+
+        return chat.ResponsePayload(
+            **{
+                "candidates": [
+                    {
+                        "message": {
+                            "role": "assistant",
+                            "content": self.process_response(resp["predictions"]),
+                        },
+                        "metadata": {},
+                    }
+                ],
+                "metadata": {
+                    "model": self.config.model.name,
+                    "route_type": self.config.route_type,
+                },
+            }
+        )
+
+    async def embeddings(self, payload: embeddings.RequestPayload) -> embeddings.ResponsePayload:
+        # Example request to MLflow REST API server for embeddings:
+        # {
+        #     "inputs": ["a sentence", "another sentence"],
+        #     "params": {
+        #         "output_value": "token_embeddings",
+        #     }
+        # }
+
+        resp = await send_request(
+            headers=self.headers,
+            base_url=self.mlflow_config.mlflow_server_url,
+            path="invocations",
+            payload=self.process_payload(payload, "text"),
+        )
+
+        # Example response:
+        # {"predictions": [[0.100, -0.234, 0.002, ...], [0.222, -0.111, 0.134, ...]]}
+
+        return embeddings.ResponsePayload(
+            **{
+                "embeddings": resp["predictions"],
```

Comment:
If the predictions don't have the expected structure, I think the resulting error will be 500 (internal error) instead of 502 (upstream server returned an invalid response). Can we validate the structure of the returned predictions first?